GVFS/GVFS.Virtualization/Projection/Readme.md

```diff
@@ -0,0 +1,181 @@
+
+### `LazyUTF8String`
+
+Class used to keep track of the string from the index that is in the `BytePool` and converts from the `BytePool` to a `string` on when needed by either calling the `GetString` method or `Compare` when one string is not all ASCII.
+
+### `SortedFolderEntries`
+
+Class used to keep the list entries for a folder, either `FolderData` or `FileData` objects) in sorted order.  This class keeps the static pool of both `FolderData` and `FileData` objects for reuse.
+
+Couple of things to note:
+
+1. This is using the `Compare` method of the `LazyUTF8String` class for a performance optimization since most of the time the paths in the index are ASCII and the code can do byte by byte comparison and __not__ have to convert to a `string` object and then compare which is a performance and memory hit.
+
+2. When getting the index of the name in the sorted entries it will return the bitwise complement of the index where the item should be inserted.  This was done to avoid making one call to determine if the name exists and a second call to get the index for insertion.
+
+### `SparseFolderData`
+
+Class used to keep the sparse folder information.  It contains a flag for whether the folder should be recursed into for projection, the depth of the folder, and the children in a name, data `Dictionary<string, SparseFolderData>`.
+
+```
+Example:
+
+For sparse folder entries of:
+foo/example
+other
+
+The SparseFolderData would be:
+root
+Children:
+|- foo (IsRecursive = false, Depth = 0)
+|  Children:
+|  |- example (IsRecursive = true, Depth = 1)
+|
+|- other (IsRecursive = true, Depth = 0)
+```
+
+### `GitIndexEntry`
+
+Class used to store the data from the index about a single entry.  There is only one instance of this class used during index parsing and it is reused for each index entry.  The reason for this is that version 4 of the git index has the [path prefix compressed](https://github.com/microsoft/git/blob/f5992bb185757a1654ce31424611b4d05bda3400/Documentation/technical/index-format.txt#L116) and the previous path is needed to create the path for the current entry.  The code in this class is heavily optimized to make parsing the index and the paths as fast as possible.
+
+### `GitIndexParser`
+
+Class that is responsible for parsing the git index based of version 4.  Please see [`index-format.txt`](https://github.com/microsoft/git/blob/f5992bb185757a1654ce31424611b4d05bda3400/Documentation/technical/index-format.txt) for detailed information of this format.  This is used to both validate the index and build the projection.  It currently ignores all index extensions and is only for getting the paths and building the tree using the `FolderData` and `FileData` classes. The index is read in chunks of 512K which gave the best performance.
+
+## Other classes
+
+### `ProjectedFileInfo`
+
+Class used to hold the data that is used by ProjFS when enumerating or creating placeholders.

```

Comment:
This got me thinking -- a good followup PR would be a discussion of the lifecycle of a file in VFSForGit.